internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -170,7 +186,40 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	}
 
 	det.SetState(details.StateExtracting)
-	// TODO: add check that no archive files end up in the current versioned home
+
+	metadata, err := u.getPackageMetadata(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("reading metadata for elastic agent version %s package %q: %w", version, archivePath, err)
+	}
+
+	currentVersion := agentVersion{
+		version:  release.Version(),
+		snapshot: release.Snapshot(),
+		hash:     release.Commit(),
```

Comment:
This doesn't use the truncated hash that we would read out of a legacy agent without a manifest file. Does that cause problems?

```go
func readCommitHash(reader io.Reader) (string, error) {
	commitBytes, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("reading agent commit hash file: %w", err)
	}
	hash := strings.TrimSpace(string(commitBytes))
	if len(hash) < hashLen {
		return "", fmt.Errorf("hash %q is shorter than minimum length %d", string(commitBytes), hashLen)
	}
	return hash, nil
}
```

Reply:
We already truncate the hash in all the relevant functions for the legacy part to 6. I preferred err on the side of caution and use the full hash in the new checks.

Reply:
Ah I see what I was missing. `metadata.hash` is not truncated anymore, it is truncated when it is used in places where that matters.

```go
	metadata, err := getPackageMetadataFromZipReader(r, fileNamePrefix)
	if err != nil {
		return UnpackResult{}, fmt.Errorf("retrieving package metadata from %q: %w", archivePath, err)
	}

	hash = metadata.hash[:hashLen]
```

I was thinking that `metadata.hash` was truncated and this comparison might evaluate to false but manual testing showed this was working properly.